skimage/segmentation/fast_draw.py

```diff
@@ -0,0 +1,278 @@
+
+from ..transform import resize as skresize
+from ..measure import label as find_label
+from ..morphology import dilation, disk
+from ..segmentation import random_walker
+from ..color import rgb2grey
+
+from .._shared.utils import warn
+
+class FastDRaW():
+    """A fast segmentation algorithm based on the random walker algorithm.
+        
+    FastDRaW implemented for 2D images as described in [1].
+    The algorithm performs in a two-step segmetnation. In the first step, a
+    random walker segmentation is performed on a small (down-sampled)
+    version of the image to obtain a coarse segmentation contour. In the
+    second step, the result is refined by applying a second random walker
+    segmentation over a narrow strip around the coarse contour.
+    
+    Parameters
+    ----------
+    image : array_like
+        Image to be segmented. If `image` is multi-channel it will
+        be converted to gray-level image before segmentation.
+    beta : float
+        Penalization coefficient for the random walker motion
+        (the greater `beta`, the more difficult the diffusion).
+    downsampled_size : int, default 100
+        The size of the down-sampled image. Should be smaller than the
+        size of the original image. Recommended values between 100 and 200.
+    tol : float
+        tolerance to achieve when solving the linear system, in
+        cg' and 'cg_mg' modes.
+    return_full_prob : bool, default False
+        If True, the probability that a pixel belongs to each of the labels
+        will be returned, instead of boolean array.
+    
+    See also
+    --------
+    skimage.segmentation.random_walker: random walker segmentation
+        The original random walker algorithm.
+        
+    References
+    ----------
+    [1] H.-E. Gueziri, L. Lakhdar, M. J. McGuffin and C. Laporte,
+    "FastDRaW - Fast Delineation by Random Walker: application to large
+    images", MICCAI Workshop on Interactive Medical Image Computing (IMIC),
+    Athens, Greece, (2016).
+    
+    Examples
+    --------
+    >>> from skimage.data import coins
+    >>> import matplotlib.pyplot as plt
+    >>> image = coins()
+    >>> labels = np.zeros_like(image)
+    >>> labels[[129, 199], [155, 155]] = 1 # label some pixels as foreground
+    >>> labels[[162, 224], [131, 184]] = 2 # label some pixels as background
+    >>> fastdraw = FastDRaW(image, beta=100, downsampled_size=100)
+    >>> segm = fastdraw.update(labels)
+    >>> plt.imshow(image,'gray')
+    >>> plt.imshow(segm, alpha=0.7)
+    """
+    
+    def __init__(self, image, beta=300, downsampled_size=100,
+                 tol=1.e-3, return_full_prob=False):
+        
+        assert (beta > 0), 'beta should be positive.'
+        self.beta = beta
+        self.return_full_prob = return_full_prob
+        self.image = rgb2grey(image)
+        ## It is important to normalize the data between [0,1] so that beta
+        ## makes sens
+        self.image = self.image / 255.0
```

Comment:
(which is nice because it doesn't rescale it if the image is already a float in [0, 1])